### PR TITLE
feat(tags): add folder-level tags with inheritance 

### PR DESCRIPTION
**Pull request by morsh-bruno**
JIRA: BRU-3344

Github: #7263, #7554, #6943 

### Description

Tags are currently supported only at the request level, requiring users to manually add the same tags across multiple requests within a folder.

This does not scale for larger workflows and introduces repetitive, error-prone setup. Folder-level tag support should allow users to define tags once and have them apply to all requests within that folder.

### Problem

Users often group related requests into folders (e.g., smoke tests, regression suites). When these requests need the same tags, users must add them individually to each request.

### Fix

The PR allows adding tags at the folder level and applies to them to all sub items under the folder. The Inherited tags are now available in Collection runs, CLI runs, doc generation and tags at Scripting.

### Screenshots

| Before | After |
| --- | --- |
| folder-settings-before | folder-settings-after |
| request-tags-before | request-tags-after |

#### Contribution Checklist:

- [x] **I've used AI significantly to create this pull request**
- [x] **The pull request only addresses one issue or adds one feature.**
- [x] **The pull request does not introduce any breaking changes**
- [x] **I have added screenshots or gifs to help explain the change if applicable.**
- [x] **I have read the [contribution guidelines](https://github.com/usebruno/bruno/blob/main/contributing.md).**
- [ ] **Create an issue and link to the pull request.**
- [x] **I've run the claude code review skill locally.**

Note: Keeping the PR small and focused helps make it easier to review and merge. If you have multiple changes you want to make, please consider submitting them as separate pull requests.

#### Publishing to New Package Managers

Please see [here](../publishing.md) for more information.

## Summary by CodeRabbit

* **New Features**
  * Added folder tags with a dedicated Folder Settings tab for viewing, adding, removing, and saving tags.
  * Folder tags cascade to nested folders and requests, with inherited tags clearly identified and expandable.
  * Prevented duplicate tags when a tag is inherited from a parent folder.
  * Runner filtering, folder runs, and CLI tag filters now support inherited tags.
  * Added folder-tag support across Bru and YAML collections, including import and export.

* **Documentation**
  * Clarified how request and folder tags are applied and managed.

### CODE REVIEW

File: packages/bruno-app/src/components/FolderSettings/Settings/Tags/index.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import { addFolderTag, deleteFolderTag } from 'providers/ReduxStore/slices/collections';
+import { saveFolderRoot } from 'providers/ReduxStore/slices/collections/actions';
+import { getInheritedTagSourcesForItem, getUniqueTagsFromItems } from 'utils/collections/index';
+import TagList from 'components/TagList/index';
+import { getFolderTags } from '@usebruno/common';
+
+const Tags = ({ folder, collection }) => {
+  const dispatch = useDispatch();
+
+  const tags = getFolderTags(folder);
```
Comment (sharan-bruno) [inline packages/bruno-app/src/components/FolderSettings/Settings/Tags/index.js:12]: u can think of better naming for this variable instead of just tags, like currentFolderTags or folderTags

**Comment by sharan-bruno** [comment]
@morsh-bruno found a couple of issues with this:

**1. Duplicate tags in nested folders**

Let's say we have a collection with Folder 1 → Folder 2.

* Add a tag in Folder 1 and try to add the same tag in Folder 2. It shows an error, which is correct.
* But if we add the tag first in Folder 2 and then try to add the same tag in Folder 1, it allows us to add it without any error.

The behaviour should be consistent in both cases.

**2. Duplicate tags when moving folders**

* Create Folder 1 and Folder 2 inside a collection.
* Add a tag called `test` in both folders.
* Drag and drop Folder 2 inside Folder 1.

It allows the drop, even though both folders have the same tag. Basically, we are preventing duplicate tags when adding them, but this edge case allows duplicate tags to be created.

**3. Flickering issue with auto-save**

* Enable auto-save with a 1000ms delay.
* Add some tags at the request level.
* Try adding and removing tags and observe the flickering issue.

The tags flicker when adding or removing them.

File: packages/bruno-app/src/components/RunnerResults/RunConfigurationPanel/index.jsx [hunk 2]
```diff
@@ -10,19 +10,18 @@ import { sortByNameThenSequence } from 'utils/common/index';
...
   if (item.type === 'ws-request' || item.type === 'grpc-request') return true;
 
   // Check tag filtering
-  const requestTags = item.draft?.tags || item.tags || [];
   const includeTags = tags?.include || [];
   const excludeTags = tags?.exclude || [];
 
   if (includeTags.length > 0 || excludeTags.length > 0) {
-    return !isRequestTagsIncluded(requestTags, includeTags, excludeTags);
+    return !isRequestTagsIncluded(getEffectiveTags(getOwnTags(item), item.inheritedTags), includeTags, excludeTags);
```
Comment (sharan-bruno) [inline packages/bruno-app/src/components/RunnerResults/RunConfigurationPanel/index.jsx]: i guess u can simplify this like, ur passing function inside, Just for readability

File: packages/bruno-app/src/utils/collections/index.js [hunk 3]
```diff
@@ -1647,32 +1660,70 @@ export const getUniqueTagsFromItems = (items = [], { includeDrafts = true } = {}
...
 
   if (tags && tags.include && tags.exclude) {
     const includeTags = tags.include ? tags.include : [];
     const excludeTags = tags.exclude ? tags.exclude : [];
-    requestItems = requestItems.filter(({ tags: requestTags = [], draft }) => {
-      requestTags = draft?.tags || requestTags || [];
-      return isRequestTagsIncluded(requestTags, includeTags, excludeTags);
-    });
+    requestItems = requestItems.filter(({ item, inheritedTags: inherited }) =>
+      isRequestTagsIncluded(getEffectiveTags(getOwnTags(item), inherited), includeTags, excludeTags)
```
Comment (sharan-bruno) [inline packages/bruno-app/src/utils/collections/index.js]: could you split the line

File: packages/bruno-app/src/components/FolderSettings/Settings/index.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import { IconTag } from '@tabler/icons';
+import Tags from './Tags/index';
+
+const Settings = ({ folder, collection }) => {
+  return (
+    <div className="h-full w-full">
+      <div className="text-xs mb-4 text-muted">Configure settings for this folder.</div>
+      <div className="bruno-form">
+        <div className="mb-6">
+          <h3 className="text-xs font-medium text-gray-900 dark:text-gray-100 flex items-center gap-1 mb-4">
```
Comment (sanish-bruno) [inline packages/bruno-app/src/components/FolderSettings/Settings/index.js]: use theme color

File: tests/folder/tags/read-folder-tags.spec.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+      await expect(locators.tags.inheritedToggle()).toBeHidden();
+    });
+  });
+
+  test('shows neither chips nor a status dot for a folder with no tags', async ({ pageWithUserData: page }) => {
+    const locators = buildCommonLocators(page);
+
+    await openFolderSettingsByPath(page, COLLECTION_NAME, ['misc']);
+
+    await expect(page.getByTestId('folder-settings-tab-settings')).toBeVisible();
```
Comment (sanish-bruno) [inline tests/folder/tags/read-folder-tags.spec.ts]: move to common-locators if possible

File: packages/bruno-cli/src/utils/collection.js [hunk 6]
```diff
@@ -488,6 +489,27 @@ const getTreePathFromCollectionToItem = (collection, _item) => {
...
+
+  const walk = (items, inheritedTags) => {
+    each(items, (item) => {
+      if (item.type === 'folder') {
+        const folderTags = getEffectiveTags(inheritedTags, getFolderTags(item));
+        tagsByPathname.set(item.pathname, folderTags);
+        walk(item.items, folderTags);
+        return;
+      }
+      tagsByPathname.set(item.pathname, getEffectiveTags(getOwnTags(item), inheritedTags));
```
Comment (sanish-bruno) [inline packages/bruno-cli/src/utils/collection.js:504]: why did we swap the order  ofarguments  inheritedTags and ownTags  for `getEffectiveTags` when it comes to folder, request types?

File: packages/bruno-electron/src/utils/collection.js [hunk 7]
```diff
@@ -852,6 +852,31 @@ const getFormattedCollectionOauth2Credentials = ({ oauth2Credentials = [] }) =>
...
+
+  const walk = (items, inheritedTags) => {
+    each(items, (item) => {
+      if (item.type === 'folder') {
+        const folderTags = getEffectiveTags(inheritedTags, getFolderTags(item));
+        tagsByUid.set(item.uid, folderTags);
+        walk(item.items, folderTags);
+        return;
+      }
+      tagsByUid.set(item.uid, getEffectiveTags(getOwnTags(item), inheritedTags));
```
Comment (sanish-bruno) [inline packages/bruno-electron/src/utils/collection.js:871]: why the did we swap the argument order for `getEffectiveTags` for folder and request

File: packages/bruno-app/src/components/RunnerResults/index.jsx [hunk 8]
```diff
@@ -97,6 +97,9 @@ export default function RunnerResults({ collection }) {
   // have tags been added for the collection run
   const areTagsAdded = tags.include.length > 0 || tags.exclude.length > 0;
 
+  // resolved for the whole tree in one walk, rather than once per result row
+  const effectiveTagsByUid = getEffectiveTagsByItemUid(collection.items);
```
Comment (sanish-bruno) [inline packages/bruno-app/src/components/RunnerResults/index.jsx]: should it be a useMemo?

File: packages/bruno-common/src/api-docs/index.ts [hunk 9]
```diff
@@ -60,7 +60,8 @@ const isRequestItem = (item: TaggedItem): boolean =>
 export const filterRequestItemsByTags = <T extends TaggedItem>(
```
Comment (sanish-bruno) [inline packages/bruno-common/src/api-docs/index.ts:60]: this function isn't draft-aware, bruno-common tag specific methods consider draft, considering the structure defined as `TaggedTreeNode`, any reason for this difference in behaviour? or all the scenarios will be handles thought `getEffectiveTags(getFolderTags(item), inheritedTags)`?

File: tests/folder/tags-yml/read-folder-tags.spec.ts [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+      await expect(locators.tags.inheritedToggle()).toBeHidden();
+    });
+  });
+
+  test('shows neither chips nor a status dot for a folder with no tags', async ({ pageWithUserData: page }) => {
+    const locators = buildCommonLocators(page);
+
+    await openFolderSettingsByPath(page, COLLECTION_NAME, ['misc']);
+
+    await expect(page.getByTestId('folder-settings-tab-settings')).toBeVisible();
```
Comment (sanish-bruno) [inline tests/folder/tags-yml/read-folder-tags.spec.ts]: use common-locators, move them if not present already

**Review by sanish-bruno — CHANGES_REQUESTED** [review]

File: packages/bruno-common/src/api-docs/index.ts (hunk 9 above)
Comment (morsh-bruno) [reply packages/bruno-common/src/api-docs/index.ts:60]: Added a new method that uses only the saved Tags for folder as well in line with the existing implementation for request tags.

File: packages/bruno-electron/src/utils/collection.js (hunk 7 above)
Comment (morsh-bruno) [reply packages/bruno-electron/src/utils/collection.js:871]: It was done so the order was consistent from outer folder to inner. But functionally, it was always a includes/excludes, maintaining the same arg order to avoid confusion

File: packages/bruno-cli/src/utils/collection.js (hunk 6 above)
Comment (morsh-bruno) [reply packages/bruno-cli/src/utils/collection.js:504]: It was done so the order was consistent from outer folder to inner. But functionally, it was always a includes/excludes, maintaining the same arg order to avoid confusion

File: packages/bruno-lang/v2/src/collectionBruToJson.js [hunk 11]
```diff
@@ -65,14 +65,19 @@ const grammar = ohm.grammar(`Bru {
...
 
   // Text Blocks
   textblock = textline (~tagend nl textline)*
   textline = textchar*
   textchar = ~nl any
 
+  // List
+  list = st* "[" nl+ listitems? st* nl+ st* "]"
+  listitems = listitem (nl+ listitem)*
+  listitem = st+ (alnum | "_" | "-")+ st*
```
[the remark this answers is not in the document]
Comment (morsh-bruno) [reply packages/bruno-lang/v2/src/collectionBruToJson.js:79]: Request level tags follow the same pattern, hence the same behavior followed here.

File: packages/bruno-app/src/components/FolderSettings/Settings/index.js (hunk 4 above)
Comment (morsh-bruno) [reply packages/bruno-app/src/components/FolderSettings/Settings/index.js]: Done. Applied on both folder and request settings

File: packages/bruno-app/src/components/RunnerResults/RunConfigurationPanel/index.jsx (hunk 2 above)
Comment (morsh-bruno) [reply packages/bruno-app/src/components/RunnerResults/RunConfigurationPanel/index.jsx]: Done

File: packages/bruno-app/src/utils/collections/index.js (hunk 3 above)
Comment (morsh-bruno) [reply packages/bruno-app/src/utils/collections/index.js]: Done.

File: packages/bruno-app/src/utils/tests/collections/folder-tags.spec.js [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+        }),
+        request({ uid: 'inner-request' })
+      ]
+    }),
+    folder({ uid: 'untagged-folder', items: [request({ uid: 'plain-request', tags: ['smoke'] })] }),
+    request({ uid: 'root-request' })
+  ]
+});
+
+const findItem = (collection, uid) => {
```
Comment (sharan-bruno) [inline packages/bruno-app/src/utils/tests/collections/folder-tags.spec.js:63]: findItem from collection or from tags, will it return matched tag or request item?

Comment (morsh-bruno) [reply packages/bruno-app/src/utils/tests/collections/folder-tags.spec.js:63]: finds the request/folder, not related to tag.

File: packages/bruno-app/src/components/FolderSettings/Settings/Tags/index.js (hunk 1 above)
Comment (morsh-bruno) [reply packages/bruno-app/src/components/FolderSettings/Settings/Tags/index.js:12]: Follows the same naming as the request sibling.

File: packages/bruno-app/src/components/RunnerResults/index.jsx (hunk 8 above)
Comment (morsh-bruno) [reply packages/bruno-app/src/components/RunnerResults/index.jsx]: Memoized.

**Comment by sharan-bruno** [comment]
@morsh-bruno changes looks fine

**Review by sanish-bruno — APPROVED** [review]
